src/xview/xcalentool/mt2ct.man
.\" $Id: mt2ct.man,v 2.2 1991/10/30 17:45:48 billr Exp $
.\"
.\" Copyright 1988, 1989, 1991 by Tektronix, Inc. - All Rights Reserved.
.\" 
.\" Permission to use, copy, modify, and distribute this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation, and that the name of Tektronix, Inc. not be used in
.\" advertising or publicity pertaining to distribution of the software
.\" without specific, written prior permission.
.\" 
.\" TEKTRONIX INCORPORATED MAKES NO REPRESENTATIONS ABOUT THE
.\" SUITABILITY OF THIS SOFTWARE FOR ANY PURPOSE.  IT IS PROVIDED "AS IS"
.\" WITHOUT EXPRESS OR IMPLIED WARRANTY.  TEKTRONIX INCORPORATED
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  IN NO
.\" EVENT SHALL TEKTRONIX INCORPORATED BE LIABLE FOR ANY SPECIAL,
.\" INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
.\" LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
.\" OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
.\" PERFORMANCE OF THIS SOFTWARE.
.\" 
.\" Author: Bill Randle, Tektronix, Inc. <billr@saab.cna.tek.com>
.\" 
.TH MT2CT 1L "27 February 1991"
.SH NAME
mt2ct, month2ct, cal2ct - appointment file conversion utilities
.SH SYNOPSIS
.B mt2ct
[
.I file
]
.br
.B month2ct
[
.I file
]
.br
.B cal2ct
[
.B \-d
.I date
]
[
.B \-r
.I days
]
[
.B \-f
.I outfile
]
[
.B \-e
]
[
.I file
]
.SH DESCRIPTION
.I Calentool
uses an appointment file format that is incompatable with several
other calendar or reminder programs.  These utilities may be used
to convert from one of those formats to the format
.I calentool
uses.
.LP
.I Mt2ct
converts files maintained by
.I monthtool(1L)
to calentool's .appointment format.
.I Month2ct
converts files maintained by
.I month(1L)
to calentool's .appointment format.
.I Cal2ct
converts files maintained by
.I calendar(1)
to calentool's .appointment format.
In all cases the converted file is written to "$HOME/.appointments".
.SH OPTIONS
An optional filename may be specified for all of the tools to convert.
The default filename depends on the specific conversion.  For
.I monthtool
it is "$HOME/.monthtool".  For
.I month
it is "$HOME/.month".  For
.I calendar
it is "$HOME/.calendar".
.LP
In addition,
.I cal2ct
accepts the following arguments:
.TP 10
.BI \-d \ date
Only convert dates on or after the day
.IR date ,
which may take on several formats:
.RS
.TP
day of current week:
Tue, Th, fr, MONDAY
.TP
numeric dates:
DD (day DD of current month, for example 27 for the 27th of this month)
.br
MO/DD (day DD of month MO, for example 11/12 for November 12)
.br
MO/DD/YY (day DD of month MO, year YY, for example 9/1/88 for 1 September 1988).
(If the
.B \-e
option is used, the date format is DD/MO and DD/MO/YY.
.TP
relative dates:
+n (n days from now, for example +1 for tomorrow)
.br
-n (n days ago, for example -1 for yesterday)
.TP
special names:
.I yesterday, today, tommorrow
.RE
.TP
.BI \-r \ days
Only convert entries for
.I days
days following the starting date.  The starting date is either today's
date or the date set via the
.B \-d
option.
If
.I days
is set to 0, only entries corresponding to
.I date
will be converted.
.TP
.B \-e
Change the format for the date set via the
.B \-d
option to expect DD/MM format rather than the normal MM/DD format.
.TP
.BI \-f \ outfile
Write the converted output into
.I outfile
rather than the default $HOME/.appointments.
.SH "SEE ALSO"
calentool(1L), monthtool(1L), month(1L), calendar(1)
.SH AUTHOR
Bill Randle (Tektronix, Inc., billr@saab.CNA.TEK.COM)
.SH BUGS
.I Monthtool
and
.I month
allow any time selections.  Because
.I calentool
only has 30 minute timeslots, all times are rounded to the nearest
30 minute interval (e.g. 1:10 becomes 1:00, 1:20 becomes 1:30 and
1:50 becomes 2:00).  All
.I calendar
file entries are converted to
.I calentool
notes, since
.I calendar
does not have a standardized way of indicating times.  They can be
manually moved to the proper timeslot using
.I calentool's
.BR cut and paste
operations.  Also, the calendar file entries must have the date at
the start of the line.
.I Cal2ct
also does not invoke the C-preprocessor on the calendar file before
processing.


src/NeWS
../NeWS

src/xview/xcalentool/mt2ct.c
/*
 * $Id: mt2ct.c,v 2.3 1994/08/19 19:53:08 billr Exp $
 */
/*
 * mt2ct - convert monthtool reminder files to calentool style files
 *
 * Copyright 1989, 1991, 1994 by Tektronix, Inc. - All Rights Reserved.
 * 
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Tektronix, Inc. not be used in
 * advertising or publicity pertaining to distribution of the software
 * without specific, written prior permission.
 * 
 * TEKTRONIX INCORPORATED MAKES NO REPRESENTATIONS ABOUT THE
 * SUITABILITY OF THIS SOFTWARE FOR ANY PURPOSE.  IT IS PROVIDED "AS IS"
 * WITHOUT EXPRESS OR IMPLIED WARRANTY.  TEKTRONIX INCORPORATED
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  IN NO
 * EVENT SHALL TEKTRONIX INCORPORATED BE LIABLE FOR ANY SPECIAL,
 * INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
 * LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 * 
 * Author: Bill Randle, Tektronix, Inc. <billr@saab.cna.tek.com>
 */ 

#include "ct.h"
#include <stdio.h>
#include <ctype.h>

struct appt_entry appts;
struct appt_entry *aptr;
char filename[128], *file;
FILE *fp;

#ifdef __STDC__
int read_mt_file (void);
void write_ct_file (void);
#else
int read_mt_file ();
void write_ct_file ();
#endif

void
main(argc, argv)
int argc;
char *argv[];
{
	if (argc > 1)
		file = argv[1];
	else {
		strcpy(filename, getenv("HOME"));
		strcat(filename, "/.monthtool");
		file = filename;
	}

	if ((fp = fopen(file, "r")) == NULL) {
		fprintf(stderr, "can't open monthtool file <%s> for reading\n", file);
		exit(1);
	}
	if (!read_mt_file()) {
		fprintf(stderr, "no reminders read from %s\n", file);
		exit(1);
	}
	fclose(fp);
	strcpy(filename, getenv("HOME"));
	strcat(filename, "/.appointments");
	if ((fp = fopen(filename, "w")) == NULL) {
		fprintf(stderr, "can't open .appointments file for writing\n");
		exit(1);
	}
	write_ct_file();
}

/*
 * read in the monthtool file, filling the appts records
 */
int
read_mt_file()
{
	char *ptr, *fgets();
	char buf[512];
	struct appt_entry *optr;

	aptr = &appts;
	optr = aptr;
	while (fgets(buf, 512, fp) != NULL) {
		aptr->flags = aptr->repeat = aptr->lookahead = 0;
		aptr->warn = 10;  /* default */
		aptr->sindex = 0;
		aptr->next = NULL;
		ptr = buf;
		while (isspace(*ptr))
			++ptr;
		aptr->month = *ptr++ - '0';
		if (isdigit(*ptr))
			aptr->month = aptr->month * 10 + (*ptr++ - '0');
		++ptr;	/* skip ',' */
		aptr->day = *ptr++ - '0';
		if (isdigit(*ptr))
			aptr->day = aptr->day * 10 + (*ptr++ - '0');
		++ptr;	/* skip ',' */
		aptr->year = 0;
		while (isdigit(*ptr))
			aptr->year = aptr->year * 10 + (*ptr++ - '0');
		if (aptr->year > 1900)
			aptr->year -= 1900;
		++ptr;	/* skip ',' */
		if (!strncmp(ptr, "    ", 4)) {
			aptr->flags |= A_NOTE;
			ptr += 5;
		} else {
			aptr->hour = (*ptr++ - '0') * 10;
			aptr->hour += *ptr++ - '0';
			aptr->minute = (*ptr++ - '0') * 10;
			aptr->minute += *ptr++ - '0';
			if (aptr->minute < 15)
				aptr->minute = 0;
			else if (aptr->minute < 45)
				aptr->minute = 30;
			else {
				aptr->minute = 0;
				aptr->hour++;
			}
			++ptr;	/* skip ',' */
		}
		strcpy(aptr->str, ptr);
		if (aptr->year == 0)
			aptr->flags |= ALL_YEARS;
		if (aptr->month == 0)
			aptr->flags |= ALL_MONTHS;
		else if (aptr->month == 99) {
			/* weekly reminder */
			aptr->flags |= ALL_MONTHS;
			aptr->flags |= Setday(aptr->day-1);
		} else
			--aptr->month;
		if (aptr->day == 0)
			aptr->flags |= ALL_DAYS;
		aptr->next = (struct appt_entry *)malloc(sizeof(struct appt_entry));
		if (aptr->next == NULL) {
			fprintf(stderr, "out of memory\n");
			return(0);
		}
		optr = aptr;
		aptr = aptr->next;
	}
	if (aptr == &appts)
		return(0);	/* nothing read */
	/* don't need the last one */
	free(aptr);
	optr->next = NULL;
	return(1);
}

/*
 * write out the new .appointments file
 */
void
write_ct_file()
{
	aptr = &appts;
	fputs(HEADER, fp);
	while (aptr) {
		if (put_aentry(fp, aptr)) {
			fprintf(stderr, "error writing .appointments file\n");
			return;
		}
		aptr = aptr->next;
	}
}
